src/Tablebot/Plugin/Discord.hs

```diff
@@ -81,6 +85,34 @@ reactToMessage m e =
   lift . restCall $
     R.CreateReaction (messageChannel m, messageId m) e
 
+-- | @getReplyMessage@ returns the message being replied to (if applicable)
+getReplyMessage :: Message -> DatabaseDiscord (Maybe Message)
+getReplyMessage m = do
+  let m' = referencedMessage m
+  let mRef = messageReference m
+  case m' of
+    Just msg -> return $ Just msg
+    Nothing -> case mRef of
+      Nothing -> return Nothing
+      Just mRef' -> maybeGetMessage (referenceChannelId mRef') (referenceMessageId mRef')
+  where
+    maybeGetMessage :: Maybe ChannelId -> Maybe MessageId -> DatabaseDiscord (Maybe Message)
+    maybeGetMessage (Just cId) (Just mId) = do
+      m' <- getMessage cId mId
+      case m' of
+        Left _ -> return Nothing
+        Right msg -> return $ Just msg
+    maybeGetMessage _ _ = return Nothing
+
+-- | @getPrecedingMessage@ returns the message immediately above the provided message
+getPrecedingMessage :: Message -> DatabaseDiscord (Maybe Message)
+getPrecedingMessage m = do
+  mlst <- lift . restCall $ R.GetChannelMessages (messageChannel m) (1, R.BeforeMessage (messageId m))
+  case mlst of
+    Right mlst' ->
+      return $ Just $ Prelude.head mlst'
```

Comment:
Sanity check: is it possible that `mlst'` will be empty? Just if so, that `head` call is concerning.